c_src/membrane_ffmpeg_swscale_plugin/pixel_format_converter/converter.c

```diff
@@ -0,0 +1,120 @@
+#include "converter.h"
+#include <libavutil/error.h>
+#include <libavutil/frame.h>
+#include <libavutil/imgutils.h>
+#include <libswscale/swscale.h>
+
+enum AVPixelFormat string_to_AVPixelFormat(char *);
+
+UNIFEX_TERM create(UnifexEnv *env, uint64_t width, uint64_t height,
+                   char *old_format, char *new_format) {
+  UNIFEX_TERM res;
+  const enum AVPixelFormat input_fmt = string_to_AVPixelFormat(old_format),
+                           output_fmt = string_to_AVPixelFormat(new_format);
+
+  if (input_fmt == AV_PIX_FMT_NONE) {
+    return create_result_error(env, "unknown_input_format");
+  } else if (!sws_isSupportedInput(input_fmt)) {
+    return create_result_error(env, "unsupported_input_format");
+  }
+
+  if (output_fmt == AV_PIX_FMT_NONE) {
+    return create_result_error(env, "unknown_output_format");
+  } else if (!sws_isSupportedOutput(output_fmt)) {
+    return create_result_error(env, "unsupported_output_format");
+  }
+
+  State *state = unifex_alloc_state(env);
+  state->sws_context =
+      sws_getContext(width, height, input_fmt, width, height, output_fmt,
+                     SWS_BICUBIC, NULL, NULL, NULL);
+
+  if (state->sws_context == NULL) {
+    res = create_result_error(env, "create_context_failed");
+    goto end;
+  }
+
+  state->width = width;
+  state->height = height;
+  state->srcFormat = input_fmt;
+  state->dstFormat = output_fmt;
+
+  // Allocate memory for both the input and output frames
+  int src_image_size =
+      av_image_alloc(state->src_data, state->src_linesize, state->width,
+                     state->height, state->srcFormat, 1);
+  state->dst_image_size =
+      av_image_alloc(state->dst_data, state->dst_linesize, state->width,
+                     state->height, state->dstFormat, 1);
+
+  if (src_image_size < 0 || state->dst_image_size < 0) {
+    sws_freeContext(state->sws_context);
+    res = create_result_error(env, "memory_allocation_failed");
+    goto end;
+  }
+
+  res = create_result_ok(env, state);
+end:
+  unifex_release_state(env, state);
+  return res;
+}
+
+UNIFEX_TERM process(UnifexEnv *env, State *state, UnifexPayload *payload) {
+  UNIFEX_TERM ret;
+
+  // copy input to av_image
+  av_image_fill_arrays(state->src_data, state->src_linesize, payload->data,
+                       state->srcFormat, state->width, state->height, 1);
```

Comment:
This is a memory leak as you override the pointers to the memory allocated at line 44